### PR TITLE
fix: gen metrics for certs with same CN diff OUs 

### PR DESCRIPTION
**Pull request by wbollock**
[ref: switch to slog for structured logging](https://github.com/aarnaud/vault-pki-exporter/commit/53ecaf36bef16130069600989223c306956f5e32)

Resolves https://github.com/aarnaud/vault-pki-exporter/issues/7

switches to slog with standard logging levels of info,debug,warn,error.
will make logging easier to include and more customizable. using
standard slog package reduces third party dependencies too

[fix: gen metrics for certs with same CN diff OUs](https://github.com/aarnaud/vault-pki-exporter/commit/d5c94f3030141360004e46b807207afb4bd08d31)

Resolves https://github.com/aarnaud/vault-pki-exporter/issues/28

Implements logic to consider both CN and OU as unique values for a
certificate in the certificate map. This way we get metrics for certs
that have the same CN but different Subject OUs - before, only CN was
considered.

This should still handle renewals correctly for the same cert, updating
metrics in-place. That is why serial number or certificate fingerprint
was not used as certs renewed in-place should retain their time series.

I debated also adding Issuer information to uniquely identify these
certs but did not think it was needed

### CODE REVIEW

**Comment by aarnaud** [comment]
Thanks @wbollock

File: cmd/main.go [hunk 1]
```diff
@@ -37,41 +39,59 @@ func init() {
...
 
+	flags.String("log-level", "info", "Set log level (options: info, warn, error, debug)")
+	if err := viper.BindPFlag("log-level", flags.Lookup("log-level")); err != nil {
+		log.Fatal("Could not set log level:", err)
+	}
+
 	flags.BoolP("prometheus", "", false, "Enable prometheus exporter, default if nothing else")
 	if err := viper.BindPFlag("prometheus", flags.Lookup("prometheus")); err != nil {
-		log.Fatal(err)
+		log.Fatal("Could not bind prometheus flag:", err)
```
[the remark this answers is not in the document]
Comment (aarnaud) [reply cmd/main.go:49]: or it can't be use  before `setLogLevel` ?

File: pkg/vault/client.go [hunk 2]
```diff
@@ -207,9 +209,9 @@ func (vault *ClientWrapper) authK8S() {
 
 	am, err := kubernetes.NewKubernetesAuth(os.Getenv("VAULT_K8S_ROLE"), kubernetes.WithMountPath(mount))
 	if err != nil {
-		log.Fatalln("Failed to create k8s auth method: ", err)
+		log.Fatal("Failed to create k8s auth method: ", err)
```
Comment (aarnaud) [inline pkg/vault/client.go:212]: slog ?

File: pkg/vault/client.go [hunk 3]
```diff
@@ -207,9 +209,9 @@ func (vault *ClientWrapper) authK8S() {
...
 	}
 	if _, err := vault.Client.Auth().Login(context.Background(), am); err != nil {
-		log.Fatalln("Failed to auth with k8s: ", err)
+		log.Fatal("Failed to auth with k8s: ", err)
```
Comment (aarnaud) [inline pkg/vault/client.go:215]: slog

File: cmd/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wbollock) [reply cmd/main.go:49]: I'll implement a little slog wrapper to make this easier

```
func slogFatal(msg string, args ...any) {
	slog.Error(msg, args...)
	os.Exit(1)
}

slogFatal("An unrecoverable error occurred", "error", "fake error")

2024/11/19 10:50:46 ERROR An unrecoverable error occurred error="fake error"
```

vs

```
log.Fatal("fake error")
2024/11/19 10:50:59 fake error
```

[the remark this answers is not in the document]
Comment (wbollock) [reply cmd/main.go]: 840797adb9d1369bc1307d0590d7ca1069fa8f1d

File: pkg/vault-mon/pki.go [hunk 4]
```diff
@@ -259,58 +257,67 @@ func (pki *PKI) loadCerts() error {
...
-					pki.certs[cert.Subject.CommonName] = cert
+				if _, exists := pki.certs[commonName]; !exists {
+					pki.certs[commonName] = make(map[string]*x509.Certificate)
 				}
 
-				if cert.NotAfter.Unix() < time.Now().Unix() {
-					pki.expiredCertsCounter++
+				// if cert is in map already or the new cert has a *later* expiration date, update map
+				// handles renewal of existing cert smoothly
+				if existingCert, ok := pki.certs[commonName][orgUnit]; !ok || existingCert.NotAfter.Before(cert.NotAfter) {
```
[the remark this answers is not in the document]
Comment (wbollock) [reply pkg/vault-mon/pki.go:314]: agreed, b82aca1ca5b0f11742ef79aabcd4075a344ec5d3

i still want to delete the revoked metrics in the prometheus run if that is okay

File: pkg/vault-mon/prometheus.go [hunk 5]
```diff
@@ -81,38 +85,54 @@ func PromWatchCerts(pkimon *PKIMon, interval time.Duration) {
...
-					expiry.WithLabelValues(certlabels...).Set(float64(cert.NotAfter.Sub(now).Seconds()))
-					age.WithLabelValues(certlabels...).Set(float64(now.Sub(cert.NotBefore).Seconds()))
-					startdate.WithLabelValues(certlabels...).Set(float64(cert.NotBefore.Unix()))
-					enddate.WithLabelValues(certlabels...).Set(float64(cert.NotAfter.Unix()))
+				for commonName, orgUnits := range pki.GetCerts() {
+					for orgUnit, cert := range orgUnits {
+
+						certlabels := getLabelValues(pkiname, cert)
+
+						if _, isRevoked := revokedCerts[cert.SerialNumber.String()]; isRevoked {
```
[the remark this answers is not in the document]
Comment (wbollock) [reply pkg/vault-mon/prometheus.go:105]: agreed, b82aca1ca5b0f11742ef79aabcd4075a344ec5d3

**Comment by wbollock** [comment]
thanks for the great review @aarnaud , updated log.Fatal to use slog and don't return revoked certs in loadCerts()

File: pkg/vault-mon/pki.go [hunk 6]
```diff
@@ -259,58 +257,83 @@ func (pki *PKI) loadCerts() error {
...
 				}
 
-				if _, ok := pki.certs[cert.Subject.CommonName]; !ok && cert.NotAfter.Unix() > time.Now().Unix() {
-					pki.certs[cert.Subject.CommonName] = cert
-					if err != nil {
-						log.Errorln(err)
+				// gather CRLs to determine revoked certs
+				revokedCerts := make(map[string]struct{})
+
+				for _, crl := range pki.GetCRLs() {
```
[the remark this answers is not in the document]
Comment (wbollock) [reply pkg/vault-mon/pki.go]: oh yes you are right! even better we can load though before we start batching, just on each loadCerts() run f0bb680cb9e5290bbd79c0d73c717fe7335ff1a2

File: pkg/logger/logger.go [hunk 7]
```diff
@@ -1,234 +1,33 @@
...
-func Fatalln(args ...interface{}) {
-	std.Fatalln(args...)
-}
-
-func checkIfTerminal(w io.Writer) bool {
-	switch v := w.(type) {
-	case *os.File:
-		return terminal.IsTerminal(int(v.Fd()))
+// Initialize default logger
+func Init(level string) {
```
[the remark this answers is not in the document]
Comment (wbollock) [reply pkg/logger/logger.go:9]: yup nice spot! cd4cb4cd1d7981dac7f365744f3e28c3bf6e5d57

**Comment by wbollock** [comment]
no problem, I'll be running it in production very soon! :laughing:

**Pull request merged**